### ISSUE TITLE
Add Menu Block features to core (feature parity with D8).

### ISSUE THREAD

**Issue by docwilmot** [backdrop/backdrop-issues#845]
See issue/recommendation from @quicksketch  at https://github.com/backdrop-ops/contrib/issues/82
Have a branch that does this at https://github.com/docwilmot/backdrop/tree/core-menu-blocks
Not quite PR ready yet I think (need to rework the config form to hide advanced features, and rename some functions), but worth a first look?

**Comment by quicksketch** [comment]
Hi @docwilmot, thanks for looking into this. Menu block module does quite a bit more than I had remembered. I think some these options are more esoteric than others. In particular, the only features I'd say were "common" would be the starting menu depth and max depth. It looks like D8 decided this as well. Screenshot:

![menu-options](https://cloud.githubusercontent.com/assets/100206/6995363/9e174768-daf7-11e4-9698-78fe4fa14b38.png)

Maybe we should trim down all the other options and leave them to contrib?

I left a few other code comments on the commit at https://github.com/docwilmot/backdrop/commit/2ec58aeba44a39921cdc95575d7a121e33755866.

**Comment by docwilmot** [comment]
Update (excuse the delay). Removed all features except level and depth. Please review.

Note that all calls to  system_menu_block_tree_output() were replaced by the pre-existing core function menu_tree_output() as the former was a copy of the latter with some extra classes and menu-specific theme overrides.

We could directly add these classes and overrides to menu_tree_output() if this is needed.

Compare:
```
function system_menu_block_tree_output(array &$tree, array $config = array()) {
  $build = array();
  $items = array();

  // Create context if no config was provided.
  if (empty($config)) {
    // Grab any menu item to find the menu_name for this tree.
    $menu_item = current($tree);
    $config['menu_name'] = $menu_item['link']['menu_name'];
  }
  $hook_menu_name = str_replace('-', '_', $config['menu_name']);

  // Pull out just the menu links we are going to render so that we
  // get an accurate count for the first/last classes.
  foreach ($tree as $key => &$value) {
    if ($tree[$key]['link']['access'] && !$tree[$key]['link']['hidden']) {
      $items[] = $tree[$key];
    }
  }

  $router_item = menu_get_item();
  $num_items = count($items);
  foreach ($items as $i => &$data) {
    $class = array();
    if ($i == 0) {
      $class[] = 'first';
    }
    if ($i == $num_items - 1) {
      $class[] = 'last';
    }
    // Set a class for the <li>-tag. Since $data['below'] may contain local
    // tasks, only set 'expanded' class if the link also has children within
    // the current menu.
    if ($data['link']['has_children'] && $data['below']) {
      $class[] = 'expanded';
    }
    elseif ($data['link']['has_children']) {
      $class[] = 'collapsed';
    }
    else {
      $class[] = 'leaf';
    }
    if (!empty($data['link']['leaf_has_children'])) {
      $class[] = 'has-children';
    }
    // Set a class if the link is in the active trail.
    if ($data['link']['in_active_trail']) {
      $class[] = 'active-trail';
      $data['link']['localized_options']['attributes']['class'][] = 'active-trail';
    }
    if ($data['link']['href'] == $_GET['q'] || ($data['link']['href'] == '<front>' && backdrop_is_front_page())) {
      $class[] = 'active';
    }
    // Set a menu link ID class.
    $class[] = 'menu-mlid-' . $data['link']['mlid'];
    // Normally, l() compares the href of every link with $_GET['q'] and sets
    // the active class accordingly. But local tasks do not appear in menu
    // trees, so if the current path is a local task, and this link is its
    // tab root, then we have to set the class manually.
    if ($data['link']['href'] == $router_item['tab_root_href'] && $data['link']['href'] != $_GET['q']) {
      $data['link']['localized_options']['attributes']['class'][] = 'active';
    }

    // Allow menu-specific theme overrides.
    $element['#theme'] = array(
      'menu_link__menu_block__' . $hook_menu_name,
      'menu_link__menu_block',
      'menu_link__' . $hook_menu_name,
      'menu_link',
    );
    $element['#attributes']['class'] = $class;
    $element['#title'] = $data['link']['title'];
    $element['#href'] = $data['link']['href'];
    $element['#localized_options'] = !empty($data['link']['localized_options']) ? $data['link']['localized_options'] : array();
    $element['#below'] = $data['below'] ? system_menu_block_tree_output($data['below'], $config) : $data['below'];
    $element['#original_link'] = $data['link'];
    // Index using the link's unique mlid.
    $build[$data['link']['mlid']] = $element;
  }
  if ($build) {
    // Make sure backdrop_render() does not re-order the links.
    $build['#sorted'] = TRUE;
    // Add the theme wrapper for outer markup.
    // Allow menu-specific theme overrides.
    $build['#theme_wrappers'][] = array(
      'menu_tree__menu_block__' . $hook_menu_name,
      'menu_tree__menu_block',
      'menu_tree__' . $hook_menu_name,
      'menu_tree',
    );
  }

  return $build;
}
```

and the core function:
```
function menu_tree_output($tree) {
  global $language_url;

  $build = array();
  $items = array();

  // Pull out just the menu links we are going to render so that we
  // get an accurate count for the first/last classes.
  foreach ($tree as $data) {
    if ($data['link']['access'] && !$data['link']['hidden']) {
      $items[] = $data;
    }
  }

  $router_item = menu_get_item();
  $num_items = count($items);
  foreach ($items as $i => $data) {
    $class = array();
    if ($i == 0) {
      $class[] = 'first';
    }
    if ($i == $num_items - 1) {
      $class[] = 'last';
    }
    // Set a class for the <li>-tag. Since $data['below'] may contain local
    // tasks, only set 'expanded' class if the link also has children within
    // the current menu.
    if ($data['link']['has_children'] && $data['below']) {
      $class[] = 'expanded';
    }
    elseif ($data['link']['has_children']) {
      $class[] = 'collapsed';
    }
    else {
      $class[] = 'leaf';
    }
    // Set a class if the link is in the active trail.
    if ($data['link']['in_active_trail']) {
      $class[] = 'active-trail';
      $data['link']['localized_options']['attributes']['class'][] = 'active-trail';
    }
    // Normally, l() compares the href of every link with $_GET['q'] and sets
    // the active class accordingly. But local tasks do not appear in menu
    // trees, so if the current path is a local task, and this link is its
    // tab root, then we have to set the class manually.
    if ($data['link']['href'] == $router_item['tab_root_href'] && $data['link']['href'] != $_GET['q']) {
      $class[] = 'active';
      $data['link']['localized_options']['attributes']['class'][] = 'active';
    }
    // Add the active class to the wrapper itself if necessary as the l()
    // function would.
    if (($data['link']['href'] == $_GET['q'] || ($data['link']['href'] == '<front>' && backdrop_is_front_page())) &&
      (empty($data['link']['language']) || $data['link']['language']->langcode == $language_url->langcode)) {
      $class[] = 'active';
    }

    // Allow menu-specific theme overrides.
    $element['#theme'] = 'menu_link__' . strtr($data['link']['menu_name'], '-', '_');
    $element['#attributes']['class'] = $class;
    $element['#title'] = $data['link']['title'];
    $element['#href'] = $data['link']['href'];
    $element['#localized_options'] = !empty($data['link']['localized_options']) ? $data['link']['localized_options'] : array();
    $element['#below'] = $data['below'] ? menu_tree_output($data['below']) : $data['below'];
    $element['#original_link'] = $data['link'];
    // Index using the link's unique mlid.
    $build[$data['link']['mlid']] = $element;
  }
  if ($build) {
    // Make sure backdrop_render() does not re-order the links.
    $build['#sorted'] = TRUE;
    // Add the theme wrapper for outer markup.
    // Allow menu-specific theme overrides.
    $build['#theme_wrappers'][] = 'menu_tree__' . strtr($data['link']['menu_name'], '-', '_');
  }

  return $build;
}
```

**Comment by quicksketch** [comment]
> We could directly add these classes and overrides to menu_tree_output() if this is needed.

I'm in favor of this instead of having two nearly identical functions.

> and menu-specific theme overrides.

I think our current menu theme overrides are sufficient. We don't need to inject "menu_block" in the theme hook names. 

Our current theme call is this:
```
    $element['#theme'] = 'menu_link__' . strtr($data['link']['menu_name'], '-', '_');
```

Which will already call `theme_menu_link_[menu_name]()` if it exists, and fallback to `theme_menu_link()` otherwise. The same sort of fallback applies to the `#theme_wrapper` call as well. So if the only change is adding the additional classes, let's add them and call it good. It also looks like we don't need to pass in the `$config` parameter at all, is that correct?

**Comment by docwilmot** [comment]
OK, all changes made, rebased and pushed. Not sure if this should have been a new PR; updated existing instead.

**Comment by docwilmot** [comment]
Pretty sure Travis laughed at my first attempt at this. 1400+ exceptions! 
Passed now though.

**Comment by klonos** [comment]
This is part of #378

**Comment by quicksketch** [comment]
Sorry @docwilmot, it's been ages since I took a look at this PR. I found a number of small issues:

- We need tests to ensure that the display of blocks is correct, testing limiting to X depth and starting level.
- On new installs, we should specify that the Primary menu block should be limited to 1 level depth. Right now we have this issue on sub-pages (Search is under Home in the menu):
![selection_027](https://cloud.githubusercontent.com/assets/100206/12084477/c49ef004-b268-11e5-9a3d-a0e1d2378fcf.png)
This may be alleviated or changed after #1457.
- The use of a static for `menu_block_set_title()` seems unnecessary. Rather than use a static, it seems like `system_menu_tree_prune_tree()` (which is the only function that calls `menu_block_set_title()`) could just return the title string, rather than using a static in another function.
- Several functions are block-specific, but their name doesn't indicate so. Maybe we could do a little renaming:
  - system_menu_default_config => system_menu_block_defaults()
  - system_menu_configure_form => system_menu_block_form()
  - system_menu_tree_build => system_menu_block_build()

**Comment by docwilmot** [comment]
This had already been reviewed by @quicksketch but its been languishing for over a year now, so thought I'd freshen it up and push it to 1.4.0.
Also added the block settings ability to custom menu blocks as well; seems a person creating a custom menu might want to have these features too, and not just system blocks as in previous plan.

**Comment by docwilmot** [comment]
To test, check out the primary navigation block; new fields for defining maximum depth of the menu, and its starting level. See sandbox.

**Comment by docwilmot** [comment]
@quicksketch a word of wisdom here please: was looking actually at #1457 again and found I needed to use `hook_form_alter()` if I needed AJAX in a block configure form, since there is no `$form_state`. 

That issue is of course to add the ability to select a "Dropdown menu style", so it would ideally have a select list of available dropdown styles in the block config form, which in turn loads the config form for that dropdown style once you select it. But AJAX cant work, so I've loaded the menu block code using `system_block_configure()` but used `system_form_alter()` to add the dropdown bit. Which makes it feel a bit tacked on to me. 

So would it be appropriate to convert all core menu blocks to use a block class and use the legacy process for the other blocks? So in `system.module`:

```
...
  $blocks['powered-by'] = array(
    'info' => t('Powered by Backdrop'),
    'description' => t('Displays a link to <a href="http://backdropcms.org">Backdrop CMS</a>.'),
    'weight' => '10',
  );
  // System-defined menu blocks. Load customizations from Menu module.
  $all_menus = function_exists('menu_get_menus') ? menu_get_menus() : array();
  foreach (menu_list_system_menus() as $menu_name => $title) {
    $custom_title = isset($all_menus[$menu_name]) ? $all_menus[$menu_name] : $title;
    $blocks[$menu_name] = array(
      'info' => check_plain(t($custom_title)),
      'description' => t('A hierarchical list of links for the @menu.', array('@menu' => t($title))),
      'class' => 'NewMenuBlockClass',    <============== THIS WOULD BE NEW!!
    );
  }
  return $blocks;
```

That way in the `NewMenuBlockClass` I can combine the menu block config with the dropdown styles and skip the `form_alter`.

The question isn't really if its doable but if we should just be doing one set of `System's` blocks one way and the other blocks another way, and if anyone feels the same about the 'tacked on' nature of using `hook_form_alter`

THis came out a lot longer and more cryptic than expected. Hope you get my point.

**Comment by quicksketch** [comment]
> So would it be appropriate to convert all core menu blocks to use a block class and use the legacy process for the other blocks? So in system.module:

Yes absolutely. And although it will probably need to wait until 2.x, we should anticipate switching menu blocks to just being a single block (e.g. keyed as just "menu") with children blocks as we did for fields. There's no reason why we should make 3 different blocks for 3 menus when they all have identical characteristics.

**Comment by quicksketch** [comment]
> This had already been reviewed by @quicksketch but its been languishing for over a year now, so thought I'd freshen it up and push it to 1.4.0.

Hi @docwilmot, I really apologize for the delay in getting this feature into core. I reviewed it again today and the code looks great! I'm going to give it another manual spin, I hoping we'll get this in for 1.4.0 yet. Thanks!

**Comment by quicksketch** [comment]
I rebased onto 1.x just to check tests against the latest 1.x code.New PR at https://github.com/backdrop/backdrop/pull/1375, identical code.

**Comment by quicksketch** [comment]
This PR looks great. I tested it out and found a few slightly confusing items, but it's not the fault of this PR:

- I quickly ran into an issue where I configured one instance of a menu block (Primary navigation) and then visited another page that had a different instance of the same block. The different configuration of the same block can be confusing.
- By default menu links are not "Shown as expanded", making it so each link has to have this option enabled in order to show up in a menu block.

Neither of these items are new, just further items for us to improve in the future.

This PR looks fantastic though, great tests and works exactly as advertised. I'm waiting for the latest PR to pass tests and then we're ready to merge.
